src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2021 Jakob Scholbach. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob Scholbach
+-/
+import field_theory.separable
+import algebra.algebra.basic
+import data.polynomial.degree
+import algebra.char_p.exp_char
+/-!
+# Separable degree
+
+This file contains basics about the separable degree of a polynomial.
+
+## Main results
+- `separable_degree`: the definition of the separable degree
+- `irreducible_has_separable_contraction`: any nonzero irreducible polynomial can be contracted
+  to a separable polynomial
+- `separable_degree_dvd_degree'`: the separable degree divides the degree, in function of the
+  exponential characteristic of the field
+- `separable_degree_dvd_degree` and `separable_degree_eq_degree` specialize the statement of
+  `separable_degree_dvd_degree`
+- `separable_degree_eq`: the separable degree is unique (for polynomials over a field)
+
+## Tags
+separable degree, degree, polynomial
+-/
+
+namespace polynomial
+
+noncomputable theory
+open_locale classical
+
+section comm_semiring
+
+variables {F : Type} [comm_semiring F]
+  (q : ℕ)
+
+/-- A separable contraction of a polynomial `f` is a separable polynomial `g` such that
+`g(x^(q^m)) = f(x)` for some `m : ℕ`.-/
+def has_separable_contraction (f : polynomial F) : Prop :=
+∃ g : polynomial F, g.separable ∧ ∃ m : ℕ, expand F (q^m) g = f
+
+variables {f : polynomial F} (hf : has_separable_contraction q f)
+
+/-- A choice of a separable contraction. -/
+def separable_contraction : polynomial F := classical.some hf
+
+/-- The separable degree of a polynomial is the degree of a given separable contraction. -/
+def separable_degree : ℕ := nat_degree (separable_contraction q hf)
+
+/-- The separable degree divides the degree, in function of the exponential characteristic of F. -/
+lemma separable_degree_dvd_degree' : ∃ m : ℕ, (separable_degree q hf) * (q ^ m) = f.nat_degree :=
+begin
+  cases (classical.some_spec hf).2 with m hm,
+  use m,
+  let g := classical.some hf,
+  have deg_g : g.nat_degree = separable_degree q hf, by refl,
+  rw [←deg_g, ←hm, nat_degree_expand (q^m) g, mul_comm],
+end
+
+/-- The separable degree divides the degree. -/
+lemma separable_degree_dvd_degree :
+  (separable_degree q hf) ∣ f.nat_degree :=
+exists.elim (separable_degree_dvd_degree' q hf) (λ a, λ ha, dvd.intro (q ^ a) ha)
```

Comment:
```suggestion
let ⟨a, ha⟩ := separable_degree_dvd_degree' q hf in dvd.intro (q ^ a) ha
```
is often more readable than using elim and `λ`s

Reply:
OK!